tests/test_prompt.py

```diff
@@ -490,3 +490,67 @@ def test_var_name_value_allowed(
     # Check answers file
     answers = yaml.safe_load((dst / ".copier-answers.yml").read_text())
     assert answers["value"] == "string"
+
+
+@pytest.mark.parametrize(
+    "choices",
+    (
+        [1, 2, 3],
+        [["one", 1], ["two", 2], ["three", 3]],
+        {"one": 1, "two": 2, "three": 3},
+    ),
+)
+def test_list_checkbox(
+    tmp_path_factory: pytest.TempPathFactory,
+    spawn: Spawn,
+    choices: Union[
+        List[int],
+        List[List[Union[str, int]]],  # actually `List[Tuple[str, int]]`
+        Dict[str, int],
+    ],
+):
+    src, dst = map(tmp_path_factory.mktemp, ("src", "dst"))
+
+    # Create template
+    build_file_tree(
+        {
+            (src / "copier.yml"): (
+                f"""\
+                _templates_suffix: {SUFFIX_TMPL}
+                _envops: {BRACKET_ENVOPS_JSON}
+                pick_one:
+                    type: list
```

Comment:
I'm not sure I like this interface.

Currently, when combining `type` and `choices`, type expresses the type of all choices. For example `type: str` tells Copier that no matter what choice is selected, it will be a string.

So I'd let that behavior as it is. For this example, we should have `type: int`.

Instead, I think it's better to add a new attribute `multiselect: true` which makes sense only when using `choices`. This way, the interface of this new option will be closer to current Copier behavior elsewhere.

WDYT?

Reply:
> Currently, when combining type and choices, type expresses the type of all choices. For example type: str tells Copier that no matter what choice is selected, it will be a string.  So I'd let that behavior as it is. For this example, we should have type: int.

I'm not sure I agree here.  Currently, with choices, you *always* get a single item back.  So it makes sense to use the type (or union of types) that you will get back.  Here, however, questionary *will* return you a `list`.  So saying `type: list` isn't a hack, it's actually describing exactly what the type of the value will be.  If you wanted to specify the type of each item, then I'd suggest using something like `list[<type>] with a more complicated cast_fn` as I mentioned in my original post. 

I'm not opposed to the more explicit `multi-select: true` flag in addition though.  But I do think you'd be setting things up for a bit of confusion if you leave `type` as the contained type, and users end up getting lists back

Reply:
Hmm makes sense. Ok, let me play a bit with it 😁

Reply:
I think this design decision is related to #750 and should take that feature proposal into consideration as well.

Reply:
With #750 in mind, I disagree with

> [...] it makes sense to use the type (or union of types) that you will get back.

because following that logic, a repeated question would need `type: list[<type>]` (in fact `type: list[list[<type>]]`) as well which would be ambiguous with multiple choice prompts.

Copier's questions syntax isn't a generic schema language like JSON Schema. It's rather a questions configuration language. I think it's important to not confuse the two. So `type` isn't the type of the variable that contains the answer to that question but rather one config setting of a question among several whose meaning depends on the kind of question defined in conjunction with the other settings.

Reply:
> Copier's questions syntax isn't a generic schema language like JSON Schema.

correct, and it looks like #750 would actually require something more like that no?  Here we're just allowing that the return value be a list of values, selected from some set of possibilities.  from my brief look at it, i think that #750, while definitely powerful and interesting, would be the one that would require the more powerful configuration language, no?

Reply:
in any case, it's more important that the maintainers here (you) like the syntax and typing plan :joy: so after you all give it some more thought, if you'd like to propose a clear alternative to the current PR, I'm happy to implement it, or feel free to close and reimplement in a follow up

Reply:
I think #750 would work perfectly fine as suggested there without a more schema-like language.

As I said, I think Copier's question configuration is not a schema language for describing an answer value or data structure. For instance, configuring

```
q:
  type: list[str]
  multiple: true
```

is redundant because `multiple: true` implies a list of answers. The same goes for multiple choice questions.

But ultimately @yajo needs to decide as he is the maintainer. I'm just trying to share my thoughts. 😉

Reply:
i agree, it's redundant, and as mentioned, i do also like the explicit `multiple:` or @yajo's `multi-select`.  The only thing that I think would be confusing is the use of `multiple` *and* the use of a non-container `type`:

```yaml
q:
  type: str
  multiple: true
```

Reply:
We currently have `yaml` and `json` types which result in getting any kind of object/list/str/float/int.. anything supported by those formats.

I think it's confusing that you have `type: list` because Copier users have used lists since long ago by using those other 2 types.

Let's just make it clear that we're letting the user do a multi-select of choices, while still letting the `type` reflect the value contained within each of those choices. Just like it happens with single-select choices. Considering all options, let's use this implementation: https://github.com/copier-org/copier/pull/1049#discussion_r1147742760

FWIW the default type is `yaml` IIRC. So we could have a `{type: yaml, multiselect: true}` question and that'd be just fine and readable. The docs should clarify usage anyways.

Reply:
@yajo I was experimenting a bit with this, before seeing this PR.

I was implementing like you said, to have a `multiple` option and keep the original prompt type - https://github.com/copier-org/copier/compare/master...brpaz:copier:multiselect?expand=1

The only issue I faced with this approach, is when rendering the jinja template, the value is passed as a string and not as a list, making any kind of operation like looping over all the choices difficult.

I managed to get a workaround, by setting the type as `json`.  And maybe `yaml` also works as you said. I think it´s not very intuitive but I can live with that. :)

So, @tlambert03, any chances that you will do these changes on this PR? Otherwise, I can open a PR from my branch, if @yajo thinks it´s in the right direction

Reply:
Go for it :)

Reply:
The truth is that the type is a bit redundant when using select (or, now, multiselect) questions.

The only purpose of a type is to convert any user input to the expected type. But when you're only selecting options, it's the template designer who wrote the possible answers. Thus, it's fair to assume that it's their responsibility to be consistent on the chosen types.

Of course, this didn't suppose a problem when it was a single answer. But, if now, it's a problem, maybe a very simple approach would be to raise an error if you use `multiple` with a type that is not `yaml` or `json`. As usual, just add docs and go for the simplest solution :)

Reply:
Thanks for the input all!  Gonna close this and let @brpaz follow up